File: typescript/packages/mpp/src/client/Charge.ts
import {
    AccountRole,
    type Address,
    address,
    appendTransactionMessageInstructions,
    type Base64EncodedWireTransaction,
    type Blockhash,
    createSolanaRpc,
    createTransactionMessage,
    getBase64EncodedWireTransaction,
    type Instruction,
    partiallySignTransactionMessageWithSigners,
    pipe,
    prependTransactionMessageInstructions,
    setTransactionMessageFeePayer,
    setTransactionMessageFeePayerSigner,
    setTransactionMessageLifetimeUsingBlockhash,
    signature as toSignature,
    signTransactionMessageWithSigners,
    type TransactionSigner,
} from '@solana/kit';
import { getSetComputeUnitLimitInstruction, getSetComputeUnitPriceInstruction } from '@solana-program/compute-budget';
import { getTransferSolInstruction } from '@solana-program/system';
import {
    findAssociatedTokenPda,
    getCreateAssociatedTokenIdempotentInstruction,
    getTransferCheckedInstruction,
} from '@solana-program/token';
import { Credential, Method } from 'mppx';

import {
    ASSOCIATED_TOKEN_PROGRAM,
    DEFAULT_RPC_URLS,
    MEMO_PROGRAM,
    resolveStablecoinMint,
    SYSTEM_PROGRAM,
    TOKEN_2022_PROGRAM,
    TOKEN_PROGRAM,
} from '../constants.js';
import * as Methods from '../Methods.js';

/**
 * Creates a Solana `charge` method for usage on the client.
 *
 * Supports two modes controlled by the `broadcast` option:
 *
 * - **Pull mode** (`broadcast: false`, default): Signs the transaction
 *   and sends the serialized bytes as a `type="transaction"` credential.
 *   The server broadcasts it to the Solana network.
 *
 * - **Push mode** (`broadcast: true`): Signs, broadcasts, confirms
 *   the transaction on-chain, and sends the signature as a `type="signature"`
 *   credential. Cannot be used with fee sponsorship.
 *
 * When the server advertises `feePayer: true` in the challenge, the client
 * sets the server's `feePayerKey` as the transaction fee payer and partially
 * signs (transfer authority only). The server adds its fee payer signature
 * before broadcasting.
 *
 * @example
 * ```ts
 * import { Mppx, solana } from 'solana-mpp-sdk/client'
 *
 * const method = solana.charge({ signer, rpcUrl: 'https://api.devnet.solana.com' })
 * const mppx = Mppx.create({ methods: [method] })
 *
 * const response = await mppx.fetch('https://api.example.com/paid-content')
 * console.log(await response.json())
 * ```
 */
export function charge(parameters: charge.Parameters) {
    const { signer, broadcast = false, onProgress } = parameters;

    const method = Method.toClient(Methods.charge, {
        async createCredential({ challenge }) {
            const { methodDetails } = challenge.request;
            const { network, feePayer: serverPaysFees } = methodDetails;

            if (serverPaysFees && broadcast) {
                throw new Error('broadcast=true cannot be used with fee sponsorship (feePayer: true)');
            }
            const encodedTx = await buildChargeTransaction({
                computeUnitLimit: parameters.computeUnitLimit,
                computeUnitPrice: parameters.computeUnitPrice,
                onProgress,
                request: challenge.request,
                rpcUrl:
                    parameters.rpcUrl ??
                    DEFAULT_RPC_URLS[network || 'mainnet-beta'] ??
                    DEFAULT_RPC_URLS['mainnet-beta'],
                signer,
            });
            const rpc = createSolanaRpc(
                parameters.rpcUrl ?? DEFAULT_RPC_URLS[network || 'mainnet-beta'] ?? DEFAULT_RPC_URLS['mainnet-beta'],
            );

            if (broadcast) {
                // ── Push mode (type="signature") ──
                onProgress?.({ type: 'paying' });

                const signature = await rpc
                    .sendTransaction(encodedTx, {
                        encoding: 'base64',
                        skipPreflight: false,
                    })
                    .send();

                onProgress?.({ signature, type: 'confirming' });
                await confirmTransaction(rpc, signature);
                onProgress?.({ signature, type: 'paid' });

                return Credential.serialize({
                    challenge,
                    payload: { signature, type: 'signature' },
                });
            }

            // ── Pull mode (type="transaction", default) ──
            onProgress?.({ transaction: encodedTx, type: 'signed' });

            return Credential.serialize({
                challenge,
                payload: { transaction: encodedTx, type: 'transaction' },
            });
        },
    });

    return method;
}

/**
 * Builds and signs the Solana transaction for an MPP charge request.
 *
 * This is the lower-level client SDK entry point for integrations that already
 * have a decoded charge request and want the transaction bytes instead of the
 * full mppx `charge()` method wrapper.
 */
export async function buildChargeTransaction(
    parameters: buildChargeTransaction.Parameters,
): Promise<Base64EncodedWireTransaction> {
    const {
        signer,
        request: { amount, currency, recipient, methodDetails },
        onProgress,
    } = parameters;
    const {
        network,
        decimals,
        tokenProgram: tokenProgramAddr,
        feePayer: serverPaysFees,
        feePayerKey,
        recentBlockhash: serverBlockhash,
        splits,
    } = methodDetails;

    // currency is "sol" for native, or the mint address for SPL tokens.
    const mint = resolveStablecoinMint(currency, network);

    const rpcUrl = parameters.rpcUrl ?? DEFAULT_RPC_URLS[network || 'mainnet-beta'] ?? DEFAULT_RPC_URLS['mainnet-beta'];
    const rpc = createSolanaRpc(rpcUrl);
    onProgress?.({
        amount,
        currency,
        feePayerKey: feePayerKey || undefined,
        recipient,
        type: 'challenge',
    });

    if (serverPaysFees && !feePayerKey) {
        throw new Error('feePayer=true requires feePayerKey in methodDetails');
    }

    const useServerFeePayer = serverPaysFees === true;

    // Compute primary amount (total minus splits).
    const splitsTotal = (splits ?? []).reduce((sum, s) => sum + BigInt(s.amount), 0n);
    const primaryAmount = BigInt(amount) - splitsTotal;
    if (primaryAmount <= 0n) {
        throw new Error('Splits consume the entire amount; primary recipient must receive a positive amount');
    }
    const hasAtaCreationSplits = splits?.some(split => split.ataCreationRequired === true) === true;
    if (!mint && hasAtaCreationSplits) {
        throw new Error('ataCreationRequired requires an SPL token charge');
    }
    if (hasAtaCreationSplits && currency !== mint) {
        throw new Error('ataCreationRequired requires currency to be an SPL token mint address');
    }

    // Build transfer instructions.
    const instructions: Instruction[] = [];
    const addMemoInstruction = (memo: string | undefined) => {
        if (!memo) return;
        const data = new TextEncoder().encode(memo);
        if (data.byteLength > 566) {
            throw new Error('memo cannot exceed 566 bytes');
        }
        instructions.push({
            accounts: [],
            data,
            programAddress: address(MEMO_PROGRAM),
        });
    };

    if (mint) {
        // ── SPL token transfers ──
        const mintAddress = address(mint);
        const tokenProg = tokenProgramAddr ? address(tokenProgramAddr) : await resolveTokenProgram(rpc, mintAddress);
        const tokenDecimals = decimals ?? 6;

        const [sourceAta] = await findAssociatedTokenPda({
            mint: mintAddress,
            owner: signer.address,
            tokenProgram: tokenProg,
        });

        const findDestinationAta = async (owner: Address) => {
            const [ata] = await findAssociatedTokenPda({
                mint: mintAddress,
                owner,
                tokenProgram: tokenProg,
            });

            return ata;
        };

        const addAtaCreation = async (owner: Address) => {
            const ata = await findDestinationAta(owner);

            if (useServerFeePayer) {
                instructions.push(
                    createAssociatedTokenAccountIdempotent(address(feePayerKey!), owner, mintAddress, ata, tokenProg),
                );
            } else {
                instructions.push(
                    getCreateAssociatedTokenIdempotentInstruction({
                        ata,
                        mint: mintAddress,
                        owner,
                        payer: signer,
                        tokenProgram: tokenProg,
                    }),
                );
            }

            return ata;
        };

        // Helper: add ATA creation + transferChecked for a recipient.
        const addSplTransfer = async (dest: string, transferAmount: bigint, createAta: boolean) => {
            const destOwner = address(dest);
            const destAta = createAta ? await addAtaCreation(destOwner) : await findDestinationAta(destOwner);

            instructions.push(
                getTransferCheckedInstruction(
                    {
                        amount: transferAmount,
                        authority: signer,
                        decimals: tokenDecimals,
                        destination: destAta,
                        mint: mintAddress,
                        source: sourceAta,
                    },
                    { programAddress: tokenProg },
                ),
            );
        };

        // Primary recipient ATA creation is intentionally out of scope.
        await addSplTransfer(recipient, primaryAmount, false);

        // Split transfers.
        for (const split of splits ?? []) {
            await addSplTransfer(
                split.recipient,
                BigInt(split.amount),
                !useServerFeePayer || split.ataCreationRequired === true,
            );
            addMemoInstruction(split.memo);
        }
    } else {
        // ── Native SOL transfers ──
        // Primary transfer to recipient.
        instructions.push(
            getTransferSolInstruction({
                amount: primaryAmount,
                destination: address(recipient),
                source: signer,
            }),
        );

        // Split transfers.
        for (const split of splits ?? []) {
            instructions.push(
                getTransferSolInstruction({
                    amount: BigInt(split.amount),
                    destination: address(split.recipient),
                    source: signer,
                }),
            );
            addMemoInstruction(split.memo);
        }
    }

    onProgress?.({ type: 'signing' });

    // Use server-provided blockhash if available, otherwise fetch one.
    const latestBlockhash = serverBlockhash
        ? {
              blockhash: serverBlockhash as Blockhash,
              lastValidBlockHeight: BigInt(0), // Server doesn't provide this; tx lifetime is managed by the blockhash itself.
          }
        : (await rpc.getLatestBlockhash().send()).value;

    const txMessage = pipe(
        createTransactionMessage({ version: 0 }),
        msg =>
            useServerFeePayer
                ? setTransactionMessageFeePayer(address(feePayerKey!), msg)
                : setTransactionMessageFeePayerSigner(signer, msg),
        msg => setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, msg),
        msg => appendTransactionMessageInstructions(instructions, msg),
        // Prepend compute budget instructions per best practice.
        msg =>
            prependTransactionMessageInstructions(
                [
                    getSetComputeUnitPriceInstruction({ microLamports: parameters.computeUnitPrice ?? 1n }),
                    getSetComputeUnitLimitInstruction({ units: parameters.computeUnitLimit ?? 200_000 }),
                ],
                msg,
            ),
    );

    // When server pays fees, partially sign (only the transfer authority).
    // The server will add its fee payer signature before broadcasting.
    const signedTx = useServerFeePayer
        ? await partiallySignTransactionMessageWithSigners(txMessage)
        : await signTransactionMessageWithSigners(txMessage);

    return getBase64EncodedWireTransaction(signedTx);
}

// ── Helpers ──

/**
 * Creates an Associated Token Account using the idempotent instruction
 * (CreateIdempotent = discriminator 1). This is a no-op if the ATA exists.
 *
 * Used in fee payer mode where the payer is the server's key (not a local
 * signer). The server adds its signature before broadcasting.
 */
function createAssociatedTokenAccountIdempotent(
    payer: Address,
    owner: Address,
    mint: Address,
    ata: Address,
    tokenProgram: Address,
): Instruction {
    return {
        accounts: [
            { address: payer, role: AccountRole.WRITABLE_SIGNER },
            { address: ata, role: AccountRole.WRITABLE },
            { address: owner, role: AccountRole.READONLY },
            { address: mint, role: AccountRole.READONLY },
            { address: address(SYSTEM_PROGRAM), role: AccountRole.READONLY },
            { address: tokenProgram, role: AccountRole.READONLY },
        ],
        data: new Uint8Array([1]),
        programAddress: address(ASSOCIATED_TOKEN_PROGRAM), // CreateIdempotent discriminator
    };
}

async function resolveTokenProgram(rpc: ReturnType<typeof createSolanaRpc>, mint: Address): Promise<Address> {
    const account = await rpc.getAccountInfo(mint, { encoding: 'base64' }).send();
    const owner = account.value?.owner;

    if (!owner) {
        throw new Error('Failed to determine token program for mint: mint account not found');
    }
    if (owner === TOKEN_PROGRAM || owner === TOKEN_2022_PROGRAM) {
        return address(owner);
    }

    throw new Error(`Failed to determine token program for mint: unexpected owner ${owner}`);
}

/**
 * Polls for transaction confirmation via getSignatureStatuses.
 * Only used in push mode.
 */
async function confirmTransaction(rpc: ReturnType<typeof createSolanaRpc>, signature: string, timeoutMs = 30_000) {
    const start = Date.now();
    while (Date.now() - start < timeoutMs) {
        const { value } = await rpc.getSignatureStatuses([toSignature(signature)]).send();
        const status = value[0];
        if (status) {
            if (status.err) {
                throw new Error(`Transaction failed: ${JSON.stringify(status.err)}`);
            }
            if (status.confirmationStatus === 'confirmed' || status.confirmationStatus === 'finalized') {
                return;
            }
        }
        await new Promise(r => setTimeout(r, 2_000));
    }
    throw new Error('Transaction confirmation timeout');
}

export declare namespace charge {
    type Parameters = {
        /**
         * If true, the client broadcasts the transaction and sends the signature
         * as a `type="signature"` credential. If false (default), the client sends
         * the signed transaction bytes as a `type="transaction"` credential and the
         * server broadcasts it.
         *
         * Cannot be used with server fee sponsorship (feePayer mode).
         */
        broadcast?: boolean;
        /** Compute unit limit. Defaults to 200,000. */
        computeUnitLimit?: number;
        /** Compute unit price in micro-lamports for priority fees. Defaults to 1. */
        computeUnitPrice?: bigint;
        /** Called at each step of the payment process. */
        onProgress?: (event: ProgressEvent) => void;
        /** Custom RPC URL. If not set, inferred from the challenge's network field. */
        rpcUrl?: string;
        /**
         * Solana transaction signer. Compatible with:
         * - ConnectorKit's `useTransactionSigner()` hook
         * - `createKeyPairSignerFromBytes()` from `@solana/kit` for headless usage
         * - Solana Keychain's `SolanaSigner` for remote signers
         * - Any `TransactionSigner` implementation
         */
        signer: TransactionSigner;
    };

    type ProgressEvent =
        | {
              amount: string;
              currency: string;
              feePayerKey?: string;
              recipient: string;
              type: 'challenge';
          }
        | { signature: string; type: 'confirming' }
        | { signature: string; type: 'paid' }
        | { transaction: string; type: 'signed' }
        | { type: 'paying' }
        | { type: 'signing' };
}

export declare namespace buildChargeTransaction {
    type Parameters = {
        /** Compute unit limit. Defaults to 200,000. */
        computeUnitLimit?: number;
        /** Compute unit price in micro-lamports for priority fees. Defaults to 1. */
        computeUnitPrice?: bigint;
        /** Called at each step of the payment build/signing process. */
        onProgress?: (
            event:
                | {
                      amount: string;
                      currency: string;
                      feePayerKey?: string;
                      recipient: string;
                      type: 'challenge';
                  }
                | { type: 'signing' },
        ) => void;
        /** Decoded request from a Solana MPP charge challenge. */
        request: {
            amount: string;
            currency: string;
            methodDetails: {
                decimals?: number;
                feePayer?: boolean;
                feePayerKey?: string;
                network?: string;
                recentBlockhash?: string;
                splits?: Array<{ amount: string; ataCreationRequired?: boolean; memo?: string; recipient: string }>;
                tokenProgram?: string;
            };
            recipient: string;
        };
        /** Custom RPC URL. If not set, inferred from the request network field. */
        rpcUrl?: string;
        /** Solana transaction signer. */
        signer: TransactionSigner;
    };
}

File: go/client/charge.go
package client

import (
	"context"
	"encoding/json"
	"strings"

	solana "github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/mpp-sdk/go"
	"github.com/solana-foundation/mpp-sdk/go/internal/solanautil"
	"github.com/solana-foundation/mpp-sdk/go/protocol"
	"github.com/solana-foundation/mpp-sdk/go/protocol/intents"
)

// BuildOptions customize client-side transaction creation.
type BuildOptions struct {
	Broadcast        bool
	ComputeUnitLimit uint32
	ComputeUnitPrice uint64
	ExternalID       string
}

// BuildChargeTransaction creates a payment credential payload from challenge fields.
func BuildChargeTransaction(
	ctx context.Context,
	signer solanautil.Signer,
	rpcClient solanautil.RPCClient,
	amount string,
	currency string,
	recipient string,
	methodDetails protocol.MethodDetails,
	options BuildOptions,
) (protocol.CredentialPayload, error) {
	total, err := parseAmount(amount)
	if err != nil {
		return protocol.CredentialPayload{}, err
	}
	primaryAmount, err := solanautil.SplitAmounts(total, methodDetails.Splits)
	if err != nil {
		return protocol.CredentialPayload{}, err
	}

	if options.ComputeUnitLimit == 0 {
		options.ComputeUnitLimit = 200_000
	}
	if options.ComputeUnitPrice == 0 {
		options.ComputeUnitPrice = 1
	}

	instructions := make([]solana.Instruction, 0, 2+2+len(methodDetails.Splits)*3)
	if ix, err := solanautil.BuildComputeUnitPrice(options.ComputeUnitPrice); err == nil {
		instructions = append(instructions, ix)
	}
	if ix, err := solanautil.BuildComputeUnitLimit(options.ComputeUnitLimit); err == nil {
		instructions = append(instructions, ix)
	}

	recipientKey, err := solana.PublicKeyFromBase58(recipient)
	if err != nil {
		return protocol.CredentialPayload{}, mpp.WrapError(mpp.ErrCodeInvalidConfig, "invalid recipient", err)
	}
	useServerFeePayer := methodDetails.FeePayer != nil && *methodDetails.FeePayer && methodDetails.FeePayerKey != "" && !options.Broadcast
	if options.Broadcast && methodDetails.FeePayer != nil && *methodDetails.FeePayer {
		return protocol.CredentialPayload{}, mpp.NewError(mpp.ErrCodeInvalidConfig, `type="signature" cannot be used with fee sponsorship`)
	}

	if isNativeSOL(currency) {
		ix, err := solanautil.BuildSOLTransfer(signer.PublicKey(), recipientKey, primaryAmount)
		if err != nil {
			return protocol.CredentialPayload{}, err
		}
		instructions = append(instructions, ix)
		if options.ExternalID != "" {
			memoIx, err := solanautil.BuildMemoInstruction(options.ExternalID)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
			instructions = append(instructions, memoIx)
		}
		for _, split := range methodDetails.Splits {
			splitKey, err := solana.PublicKeyFromBase58(split.Recipient)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
			splitAmount, err := parseAmount(split.Amount)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
			ix, err := solanautil.BuildSOLTransfer(signer.PublicKey(), splitKey, splitAmount)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
			instructions = append(instructions, ix)
			if split.Memo != "" {
				memoIx, err := solanautil.BuildMemoInstruction(split.Memo)
				if err != nil {
					return protocol.CredentialPayload{}, err
				}
				instructions = append(instructions, memoIx)
			}
		}
	} else {
		resolvedMint := protocol.ResolveMint(currency, methodDetails.Network)
		mint, err := solana.PublicKeyFromBase58(resolvedMint)
		if err != nil {
			return protocol.CredentialPayload{}, err
		}
		tokenProgram, err := solanautil.ResolveTokenProgram(ctx, rpcClient, mint, methodDetails.TokenProgram)
		if err != nil {
			return protocol.CredentialPayload{}, mpp.WrapError(mpp.ErrCodeRPC, "resolve token program", err)
		}
		decimals := uint8(6)
		if methodDetails.Decimals != nil {
			decimals = *methodDetails.Decimals
		}
		sourceATA, err := solanautil.FindAssociatedTokenAddressWithProgram(signer.PublicKey(), mint, tokenProgram)
		if err != nil {
			return protocol.CredentialPayload{}, err
		}
		payer := signer.PublicKey()
		if useServerFeePayer {
			payer, err = solana.PublicKeyFromBase58(methodDetails.FeePayerKey)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
		}
		addTransfer := func(owner solana.PublicKey, amount uint64) error {
			destATA, err := solanautil.FindAssociatedTokenAddressWithProgram(owner, mint, tokenProgram)
			if err != nil {
				return err
			}
			createATA, err := solanautil.BuildCreateAssociatedTokenAccount(payer, owner, mint, tokenProgram)
			if err != nil {
				return err
			}
			transfer, err := solanautil.BuildTransferChecked(amount, decimals, sourceATA, mint, destATA, signer.PublicKey(), tokenProgram)
			if err != nil {
				return err
			}
			instructions = append(instructions, createATA, transfer)
			return nil
		}
		if err := addTransfer(recipientKey, primaryAmount); err != nil {
			return protocol.CredentialPayload{}, err
		}
		if options.ExternalID != "" {
			memoIx, err := solanautil.BuildMemoInstruction(options.ExternalID)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
			instructions = append(instructions, memoIx)
		}
		for _, split := range methodDetails.Splits {
			splitKey, err := solana.PublicKeyFromBase58(split.Recipient)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
			splitAmount, err := parseAmount(split.Amount)
			if err != nil {
				return protocol.CredentialPayload{}, err
			}
			if err := addTransfer(splitKey, splitAmount); err != nil {
				return protocol.CredentialPayload{}, err
			}
			if split.Memo != "" {
				memoIx, err := solanautil.BuildMemoInstruction(split.Memo)
				if err != nil {
					return protocol.CredentialPayload{}, err
				}
				instructions = append(instructions, memoIx)
			}
		}
	}

	blockhash, err := solanautil.ResolveRecentBlockhash(ctx, rpcClient, methodDetails.RecentBlockhash)
	if err != nil {
		return protocol.CredentialPayload{}, mpp.WrapError(mpp.ErrCodeRPC, "fetch recent blockhash", err)
	}
	payer := signer.PublicKey()
	txOpts := []solana.TransactionOption{}
	if useServerFeePayer {
		payer, err = solana.PublicKeyFromBase58(methodDetails.FeePayerKey)
		if err != nil {
			return protocol.CredentialPayload{}, err
		}
	}
	txOpts = append(txOpts, solana.TransactionPayer(payer))
	tx, err := solana.NewTransaction(instructions, blockhash, txOpts...)
	if err != nil {
		return protocol.CredentialPayload{}, err
	}
	if err := solanautil.SignTransaction(tx, signer); err != nil {
		return protocol.CredentialPayload{}, err
	}

	if options.Broadcast {
		signature, err := solanautil.SendTransaction(ctx, rpcClient, tx)
		if err != nil {
			return protocol.CredentialPayload{}, mpp.WrapError(mpp.ErrCodeRPC, "send transaction", err)
		}
		if err := solanautil.WaitForConfirmation(ctx, rpcClient, signature); err != nil {
			return protocol.CredentialPayload{}, mpp.WrapError(mpp.ErrCodeTransactionFailed, "confirm transaction", err)
		}
		return protocol.CredentialPayload{Type: "signature", Signature: signature.String()}, nil
	}

	encoded, err := solanautil.EncodeTransactionBase64(tx)
	if err != nil {
		return protocol.CredentialPayload{}, err
	}
	return protocol.CredentialPayload{Type: "transaction", Transaction: encoded}, nil
}

// BuildCredentialHeader creates an Authorization header from a challenge.
func BuildCredentialHeader(
	ctx context.Context,
	signer solanautil.Signer,
	rpcClient solanautil.RPCClient,
	challenge mpp.PaymentChallenge,
) (string, error) {
	return BuildCredentialHeaderWithOptions(ctx, signer, rpcClient, challenge, BuildOptions{})
}

// BuildCredentialHeaderWithOptions creates an Authorization header from a challenge.
func BuildCredentialHeaderWithOptions(
	ctx context.Context,
	signer solanautil.Signer,
	rpcClient solanautil.RPCClient,
	challenge mpp.PaymentChallenge,
	options BuildOptions,
) (string, error) {
	var request intents.ChargeRequest
	if err := challenge.Request.Decode(&request); err != nil {
		return "", err
	}
	var details protocol.MethodDetails
	if request.MethodDetails != nil {
		raw, err := json.Marshal(request.MethodDetails)
		if err != nil {
			return "", err
		}
		if err := json.Unmarshal(raw, &details); err != nil {
			return "", err
		}
	}
	options.ExternalID = request.ExternalID
	payload, err := BuildChargeTransaction(ctx, signer, rpcClient, request.Amount, request.Currency, request.Recipient, details, options)
	if err != nil {
		return "", err
	}
	credential, err := mpp.NewPaymentCredential(challenge.ToEcho(), payload)
	if err != nil {
		return "", err
	}
	return mpp.FormatAuthorization(credential)
}

func parseAmount(value string) (uint64, error) {
	request := intents.ChargeRequest{Amount: strings.TrimSpace(value)}
	return request.ParseAmount()
}

func isNativeSOL(currency string) bool {
	return strings.EqualFold(currency, "sol")
}
